src/tests/java/htsjdk/samtools/SAMFileWriterFactoryTest.java

```diff
@@ -141,10 +144,123 @@ private void createSmallBamToOutputStream(final OutputStream outputStream,boolea
         writer.close();
     }
    
-   private void fillSmallBam(SAMFileWriter writer) {
+   private int fillSmallBam(SAMFileWriter writer) {
        final SAMRecordSetBuilder builder = new SAMRecordSetBuilder();
        builder.addUnmappedFragment("HiMom!");
-       for (final SAMRecord rec: builder.getRecords()) writer.addAlignment(rec);
-   }    
-   
+       int numRecs = builder.getRecords().size();
+       for (final SAMRecord rec: builder.getRecords()) {
+           writer.addAlignment(rec);
+       }
+       return numRecs;
+    }
+
+    private File prepareOutputFile(String extension) throws IOException {
+        final File outputFile = File.createTempFile("tmp.", extension);
+        outputFile.delete();
+        outputFile.deleteOnExit();
+        return outputFile;
+    }
+
+    //  Create a writer factory that creates and index and md5 file and set the header to coord sorted
+    private SAMFileWriterFactory createWriterFactoryWithOptions(SAMFileHeader header) {
+        final SAMFileWriterFactory factory = new SAMFileWriterFactory();
+        factory.setCreateIndex(true);
+        factory.setCreateMd5File(true);
+        // index only created if coordinate sorted
+        header.setSortOrder(SAMFileHeader.SortOrder.coordinate);
+        header.addSequence(new SAMSequenceRecord("chr1", 123));
+        header.addReadGroup(new SAMReadGroupRecord("1"));
+        return factory;
+    }
+
+    private void verifyWriterOutput(File outputFile, ReferenceSource refSource, int nRecs, boolean verifySupplementalFiles) {
+        if (true == verifySupplementalFiles) {
+            final File indexFile = SamFiles.findIndex(outputFile);
+            indexFile.deleteOnExit();
+            final File md5File = new File(outputFile.getParent(), outputFile.getName() + ".md5");
+            md5File.deleteOnExit();
+            Assert.assertTrue(indexFile.length() > 0);
+            Assert.assertTrue(md5File.length() > 0);
+        }
+
+        SamReaderFactory factory =  SamReaderFactory.makeDefault().validationStringency(ValidationStringency.LENIENT);
+        if (refSource != null) {
+            factory.referenceSource(refSource);
+        }
+        SamReader reader = factory.open(outputFile);
+        SAMRecordIterator it = reader.iterator();
+        int count = 0;
+        for (; it.hasNext(); it.next()) {
+            count++;
+        }
+
+        Assert.assertTrue(count == nRecs);
+    }
+
+    @DataProvider(name="bamOrCramWriter")
+    public Object[][] bamOrCramWriter() {
+        return new Object[][] {
+                { BamFileIoUtils.BAM_FILE_EXTENSION, },
+                { CramIO.CRAM_FILE_EXTENSION }
+        };
+    }
+
+    @Test(dataProvider="bamOrCramWriter")
+    public void testMakeWriter(String extension) throws Exception {
```

Comment:
these 4 tests look like they should share 1 a common method to reduce duplication